Add missing sequence point in lowering of using statement

jcouv opened the pull request:
**Customer scenario**
Write a conditional in a using statement without braces. When debugging, it will appear as if the second branch of the conditional executed too. 

![image](https://cloud.githubusercontent.com/assets/41759/25209772/f6d34588-2531-11e7-8e65-e35e542236ec.png)

**Bugs this fixes:**
Fixes https://github.com/dotnet/roslyn/issues/18844
Fixes https://devdiv.visualstudio.com/defaultcollection/DevDiv/_workitems?_a=edit&id=390881

**Workarounds, if any**
Add braces or don't get spooked by the misleading debugging experience.

**Risk**
**Performance impact**
Low. Just adding a hidden sequence point.

**How was the bug found?**
Reported by another Microsoft team.

@tmat @dotnet/roslyn-compiler for review.

jcouv commented:
I'm going to do some exploratory testing and code investigation. I suspect this bug exists in other scenarios where lowering introduces a try/catch.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs:
```diff
@@ -341,10 +341,13 @@ private BoundStatement RewriteUsingStatementTryFinally(SyntaxNode syntax, BoundB
             }
 
+            // Terminate the try block with a hidden sequence point
+            BoundBlock terminatedTryBlock = tryBlock.WithStatements(tryBlock.Statements.Add(new BoundSequencePoint(null, null)));
```
>BoundBlock terminatedTryBlock = tryBlock.WithStatements(tryBlock.Statements.Add(new BoundSequencePoint(null, null))); [](start = 12, length = 117)

It feels like the problem could actually be with ```if``` statements and it might manifest itself in other contexts. I would try to see if it makes more sense to tie the fix to the ```if``` lowering instead.

jcouv replied:
Yes, I wonder about that.

I can find the same problem with `while` and `for` loops:
```C#
    using (var stream = new System.IO.MemoryStream())
            while (x)
                x = false;
```
```C#
   using (var stream = new System.IO.MemoryStream())
            for ( ; x == true; )
                x = false;
```

On the other hand, I suspect there are other scenarios that introduce a `try/catch`. So far, we fixed `async` and `using`, but I think there are more. I'll try to confirm if more scenarios are broken.

jcouv replied:
Ignore that. When I try those scenarios in the debugger, they are actually fine.

jcouv replied:
I'm able to repro the problem in the debugger for this scenario:
```C#
        void F3(bool x)
        {
            string y = "";
            lock (y)
                if (!x)
                    System.Console.Write(1);
                else
                    System.Console.Write(2);             
        }
```
(lowering lock introduces a try/catch)

jcouv replied:
The `fixed` statement also introduced a try/catch, but it adds a hidden sequence point (terminating the body) already.

AlekseyTs commented:
Done with review pass.

tmat commented:
:shipit:

jcouv commented:
I found that `lock` had a similar problem, which resulted in three specific fixes (`async`, `using` and `lock`).
So I reverted the change from `async` fix (https://github.com/dotnet/roslyn/pull/16803) and moved the fix to the lowering of `if` statement instead. This single fix now handles all identified cases.
Sorry for changing the PR after it was reviewed.

jcouv commented:
@tmat @VSadov @AlekseyTs Could you take another look?
The sequence point is not inserted when lowering the conditional. That catches all the identified scenarios.

jcouv commented:
@tmat @VSadov @AlekseyTs Could you take another look?
The sequence point is not inserted when lowering the conditional. That catches all the identified scenarios.

tmat commented on src/Compilers/CSharp/Test/Emit/PDB/PDBAsyncTests.cs:
```diff
@@ -159,7 +159,7 @@ .maxstack  3
     IL_00ea:  callvirt   ""bool System.Threading.EventWaitHandle.Set()""
     IL_00ef:  pop
-   ~IL_00f0:  leave.s    IL_010c
+    IL_00f0:  leave.s    IL_010c
```
Is removal of this hidden sequence point it intentional (all the async/interator tests below seem to be missing the same sequence point)?

jcouv replied:
Yes. This sequence point was added by my original fix (https://github.com/dotnet/roslyn/pull/16803/files, which only handled conditional in async method). It it not essential because the only way to get there is through the code right above (there are no conditional jumps to the `leave.s`).

tmat replied:
OK.

VSadov replied:
Right, and this seems more robust solution since ideally we should put the hidden SP not on the goto, but on the landing/label IL that is not otherwise associated with any code.

tmat commented:
👍

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IfStatement.cs:
```diff
@@ -60,6 +60,7 @@ public override BoundNode VisitIfStatement(BoundIfStatement node)
                 builder.Add(new BoundConditionalGoto(rewrittenCondition.Syntax, rewrittenCondition, false, afterif));
                 builder.Add(rewrittenConsequence);
                 builder.Add(new BoundLabelStatement(syntax, afterif));
+                builder.Add(new BoundSequencePoint(null, null));
```
>builder.Add(new BoundSequencePoint(null, null)); [](start = 16, length = 48)

Should the sequence point go before the label? Or, perhaps, on it ```new BoundSequencePoint(null, new BoundLabelStatement(syntax, afterif))```?

jcouv replied:
Thanks. That does make more sense. I moved the seq point before the label.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IfStatement.cs:
```diff
@@ -87,6 +88,7 @@ public override BoundNode VisitIfStatement(BoundIfStatement node)
                 builder.Add(new BoundLabelStatement(syntax, alt));
                 builder.Add(rewrittenAlternativeOpt);
                 builder.Add(new BoundLabelStatement(syntax, afterif));
+                builder.Add(new BoundSequencePoint(null, null));
```
>builder.Add(new BoundSequencePoint(null, null)); [](start = 16, length = 48)

Same comment.

jcouv commented:
@dotnet-bot test windows_release_vs-integration_prtest please

jcouv commented:
@dotnet/roslyn-infrastructure FYI, restarted VS integration tests due to BasicExpressionEvaluator.StateMachineTypeParameters failure ([details](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/4323/))

jcouv commented:
@dotnet-bot test windows_release_vs-integration_prtest please

jcouv commented:
@dotnet-bot test windows_release_vs-integration_prtest please

jcouv commented:
@VSadov do you want to take another look as well?

jcouv commented:
@MeiChin-Tsai for ask-mode approval. Description up-top.

MeiChin-Tsai commented:
approved.
